File: java/24.两两交换链表中的节点.java
/*
 * @lc app=leetcode.cn id=24 lang=java
 *
 * [24] 两两交换链表中的节点
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    /**
     * 递归解法
     * 0ms 100% 30%  36.6MB
     */
    public ListNode swapPairs1(ListNode head) {
        if(head == null || head.next == null) return head;
        ListNode temp = head.next.next;
        ListNode tempHead = head.next;
        head.next.next = head;
        head.next=swapPairs(temp);
        return tempHead;

    }

    /**
     * 迭代解法
     * 0ms 100%  41%  36.6MB
     */
    public ListNode swapPairs(ListNode head) {
        if(head == null || head.next == null)return head;
        ListNode newHead=head.next;
        ListNode pre = head;
        while(head != null && head.next!=null){
            ListNode tmp = head.next.next;
            head.next.next = head;
            // if(newHead==null) newHead = head.next;
            pre.next = head.next;
            head.next=tmp;
            pre = head;
            head = tmp;
        }
        return newHead;

    }
    
}
// @lc code=end

File: py/24.两两交换链表中的节点.py
#
# @lc app=leetcode.cn id=24 lang=python3
#
# [24] 两两交换链表中的节点
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    #      递归解法
    # 40ms 75%  26%  13.4MB
    def swapPairs1(self, head: ListNode) -> ListNode:
        if head is None or head.next is None:
            return head
        tmp = head.next.next
        head.next.next = head
        res = head.next
        head.next = self.swapPairs(tmp)
        return res
    
    # 迭代解法
    # 40ms   75%    6%    13.5MB
    def swapPairs(self, head: ListNode) -> ListNode:
        if head is None or head.next is None:
            return head
        pre,res = head,head.next
        while head is not None and head.next is not None:
            tmp = head.next.next
            pre.next = head.next
            head.next.next = head
            pre = head
            head.next = tmp
            head = tmp
        return res

# @lc code=end

